### PR TITLE
Add sound implementation

### PR DESCRIPTION
**Pull request by kasamikona**
Sound emulation is functional for all 11 games as far as the title screen and sound effects.
PC Collection has other bugs causing sound effects to hang or music to end early. Wanwan lacks some sound effects as the Oki PCM expansion is not implemented yet.
The built-in demo music used by some games on the printing screen also doesn't play, currently a very minor issue.
Based entirely on my own uPD937 reverse engineering work.

### CODE REVIEW

**Comment by kasamikona** [comment]
As per initial feedback, using callbacks instead of direct calls for serial tx connection (33ab7ff), and moving SDL audio code into the sound subproject (next commit).

**Comment by kasamikona** [comment]
SDL audio code moved out of main.cpp. Now main only loads sound ROM and calls Sound::set_mute from window events in main loop. Several other minor fixes/changes inside and outside audio code.

File: src/sound/loopysound.cpp [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+		}
+	}
+	// TODO reset state?
+}
+
+void UPD937_Core::setVolumeSlider(int group, int slider) {
+	if(slider < 0) slider = 0;
+	if(slider > 4) slider = 4;
+	if(group < 0) group = 0;
+	if(group > 1) group = 1;
```
Comment (PSI-Rockin) [inline src/sound/loopysound.cpp:125]: Use `std::clamp` for both pairs here.

**Review by PSI-Rockin — CHANGES_REQUESTED** [review]

File: src/video/video.cpp [hunk 2]
```diff
@@ -102,6 +102,27 @@ static void dump_all_bmps()
...
+		{
+			auto irq_id = SH2::OCPM::INTC::IRQ::IRQ0;
+			SH2::OCPM::INTC::assert_irq(irq_id, 0);
+			SH2::OCPM::INTC::deassert_irq(irq_id);
+		}
+	}
+	//IRQ1 is triggered on visible lines when in HSYNC mode
+	if (vdp.sync_irq_ctrl.irq1_enable && (vdp.sync_irq_ctrl.irq1_source == 1))
+	{
+		if(vdp.vcount < 0x100)
```
[the remark this answers is not in the document]
Comment (kasamikona) [reply src/video/video.cpp:119]: Supposed to be checking the high bit to see if it's in the "negative" vblank area but comparison against visible_scanlines would also work.

**Comment by kasamikona** [comment]
I didn't really intend to make style changes within the loopysound area, I can do that after everything else if necessary but will be an added delay as I mostly just made syntax changes when porting. I have kept the style consistent outside of that.

The vcount magic number comparison was supposed to be checking the high bit for vblank state but comparing against visible_scanlines should achieve the same thing.

Everything else I agree with, I'll figure out how to do smart pointer stuff and get all those cases fixed as well as clamp (again mostly just syntax changes from the port which had no clamp function).

**Comment by PSI-Rockin** [comment]
We can save the style changes within the LoopySound code for another time.

As for the VCOUNT comparison, since the intent is to only fire the IRQ on visible scanlines, I think the code becomes clearer by using `vdp.visible_scanlines`. Of course, checking bit 8 achieves the same thing... I just wanted the magic number to be replaced.

**Comment by kasamikona** [comment]
Finished all requested changes including style

File: src/sound/loopysound.cpp [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+
+	// Hard correction, keep within sane distance of local time
+	if(time_reference_samples < out_sample_count)
+	{
+		time_reference_samples = out_sample_count;
+	}
+	else if(time_reference_samples > out_sample_count + (2 * buffer_size))
+	{
+		time_reference_samples = out_sample_count + (2 * buffer_size);
+	}
```
Comment (PSI-Rockin) [inline src/sound/loopysound.cpp:716]: Use `std::clamp` here.

File: src/sound/loopysound.cpp [hunk 4]
```diff
@@ -0,0 +1,901 @@
...
+	update_sample();
+	for(int lr = 0; lr <= 1; lr++)
+	{
+		int accum = 0;
+		for(int v = 0; v < 32; v += 2)
+		{
+			UPD937_VoiceState *vo = &voices[v+lr];
+			UPD937_ChannelState *ch = &channels[vo->channel];
+			if(vo->volume == 0) continue;
+			if(ch->mute) continue;
```
Comment (PSI-Rockin) [inline src/sound/loopysound.cpp:79]: Probably better to combine these two if-statements?

File: src/sound/loopysound.cpp [hunk 5]
```diff
@@ -0,0 +1,901 @@
...
+			//if(pitch_relative < 0) pitch_relative = 0;
+			//if(pitch_relative > 0x5FF) pitch_relative = 0x5FF;
+			vo->sample_fract += read_rom_16(ptr_pitchtable + pitch_relative*2);
+			if(vo->sample_fract >= 0x8000)
+			{
+				vo->sample_fract -= 0x8000;
+				vo->sample_last_val = (read_rom_16(vo->sample_ptr*2)>>4) - 0x800;
+				vo->sample_ptr++;
+			}
+			if(vo->sample_ptr > vo->sample_end) vo->sample_ptr = vo->sample_loop;
```
Comment (PSI-Rockin) [inline src/sound/loopysound.cpp:297]: Just checking - should this be `>=` instead? I would expect the end address to be exclusive, but perhaps that's not the case here

**Review by PSI-Rockin — CHANGES_REQUESTED** [review]

File: src/sound/loopysound.cpp (hunk 5 above)
Comment (kasamikona) [reply src/sound/loopysound.cpp:297]: End address is inclusive, testing with exclusive is horribly detuned as expected.

**Pull request merged**